tests/run-tests-parallel.py

```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+# Copyright (c) 2014-2021, The Monero Project
+#
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without modification, are
+# permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this list of
+#    conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice, this list
+#    of conditions and the following disclaimer in the documentation and/or other
+#    materials provided with the distribution.
+#
+# 3. Neither the name of the copyright holder nor the names of its contributors may be
+#    used to endorse or promote products derived from this software without specific
+#    prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+# THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+# THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+# The parallelism celing can be controlled with MONERO_PARALLEL_TEST_JOBS env variable,
+# for example:
+#
+# MONERO_PARALLEL_TEST_JOBS=1     tests/run-tests-parallel.py
+# MONERO_PARALLEL_TEST_JOBS=nproc tests/run-tests-parallel.py
+#
+# The minimum between the ceiling and a currently selected reasonable number of threads is used in the end.
+# The reasonable number is selected as a number, that delivers the solution in the shortest time.
+"""
+
+import os
+import sys
+import subprocess
+import multiprocessing
+from timeit import default_timer as timer
+
+NUM_PROC_MAX_REASONABLE=2 # This might be increased, once the core_tests are divided into many more independent pieces or simply sped up
+TESTS_EXCLUDED_REGEX="unit_tests|functional_tests_rpc" # These tests collide with core_tests
+ERR_CODE=1
+
+def get_forced_ceiling():
+    key = 'MONERO_PARALLEL_TEST_JOBS'
+    if key not in os.environ:
```

Comment:
A bit easier on the eyes, and saves a dict lookup, to just fetch the key and catch the exception:
```
try:
    ceiling = os.environ['MONERO_PARALLEL_TEST_JOBS']
except KeyError:
    ceiling = multiprocessing.cpu_count()
else:
    if ceiling == 'nproc':
        ceiling = multiprocessing.cpu_count()
    else:
        ceiling = int(ceiling)
```

Reply:
Makes sense. I always have (this)[https://softwareengineering.stackexchange.com/questions/189222/are-exceptions-as-control-flow-considered-a-serious-antipattern-if-so-why] mind.

Reply:
Two things:
1. I find it best not to require the reviewer to read forum links and similar (I'm not much inclined to visit stackexchange), and instead briefly state whatever may be contained therein that you find relevant.
2. As I can glean the contents of the argument from the URL, I will simply state that most arguments of the form "x pattern is bad" tend to be irrelevant on their own: In general, the pattern exists because it can be useful; like all tools, when improperly used, it can be harmful. So, the question is, whether the pattern proves reasonable in a particular context. Here, the exception-based version is quite readable, as it's a standard python idiom, and the overhead isn't a concern, but, if you don't like exceptions, then you could instead use `.get`.

Reply:
The discussion under the link suggests, that: 
1) In general it's not a good approach to use exceptions for control the logic
2) In Python it's actually expected to use them this way. Most probably, because the runtime overhead of exceptions in case of Python is much lower than in other languages

So my thinking is, that we shouldn't do this for the core implementation, but I will leave it like it is for Python. Providing a link brings more context and more independent voices. I hope that my 2 above points would describe the content under the link well enough.

Anyway, for me the PR is ready, and if you have no more comments, I'd ask you to approve.
Cheers.